doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,979 @@
+          errs() << "I've found one sir!\n";
+        }
+      }
+    }
+    return false;
+   }
+
+  Constant *isValidCandidateOperand(Value *V) {
+    if (Constant *C = dyn_cast<Constant>(V)) {
+      // Checking constant eligibility
+      if (isa<PointerType>(C->getType())) {
+        // dbgs() << "Ignoring NULL pointers\n";
+        return nullptr;
+      } else if (C->getType()->isFloatingPointTy()) {
+        // dbgs() << "Ignoring Floats 0\n";
+        return nullptr;
+      } else if (C->isNullValue()) {
+        return C;
+      } else {
+        return nullptr;
+      }
+    } else {
+      return nullptr;
+    }
+  }
+
+The ``runOnBasicBlock`` method is going to iterate through all the instructions of the block (``for`` loop) and check if any operand of those instructions is an eligible null literal.
+If any of the operand is a null literal we print a message on the debug stream and we continue.
+You may have noticed the for loop is initialized with ``BB.getFirstInsertionPt()``.
+We could have iterated through the block with a foreach like:
+
+.. code:: C++
+
+    for(auto &I : BB) {
+    }
+
+But we do not want to modify some of the special instructions located at the begining of the block (the `phi instructions <http://en.wikipedia.org/wiki/Static_single_assignment_form#Converting_out_of_SSA_form>`_), so we skip them altogether and set the iterator to the first 'normal' instruction.
+
+The ``isValidCandidateOperand`` method checks if its parameter is a literal (constant means literal in LLVM, not variable declared ``const``).
+It also checks the type of the literal, it must not be a pointer or a floating point value (you will see later why).
+The type checks are done with the ``isa<>`` function which checks if its parameter can be cast to the type given by the template parameter.
+(isa<> is used in LLVM instead of RTTI(run time type information) because it was deemed too `expensive <http://llvm.org/docs/CodingStandards.html#do-not-use-rtti-or-exceptions>`_.)
+If all those conditions are satisfied and the litteral is null we return a pointer to the operand (cast as a Constant) else ``nullptr``.
+
+If you compile and run the pass on our test code it finds **two** null literals when we just expected it to find the ``return 0``.
+
+Let's take a look at the LLVM bytecode generated by clang:
+
+.. code:: bash
+
+    # The pass is not necessary now since it doesn't change anything, but it will be later.
+    >$ clang++ -S -emit-llvm -Xclang -load -Xclang $PASSDIR/build/llvm-passes/LLVMMyPass.so path/to/test/file -o /tmp/awesome.ll
+
+We get the following:
+
+.. code:: llvm
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @foo() #0 {
+      ret i32 1
+    }
+
+    ; Function Attrs: nounwind uwtable
+    define i32 @main() #0 {
+      %1 = alloca i32, align 4 ; This instruction...
+      store i32 0, i32* %1     ; ... and this one are useless, they would be deleted if we used an optimization flag.
```

Comment:
They are not useless taken separately, they produce the identity and can be thrown out when they come together (if the other uses are not too complex)